proposals/p1885.md

```diff
@@ -657,6 +701,91 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
 These two options could allow interoperability with a compatible C++ type,
 either implicitly (template approach) or explicitly (adapter approach).
 
+#### Iterating over Carbon types in C++
+
+C++ needs to provide `begin()` and `end()` methods or free functions, in
+addition to `*` and `++` operators.
+
+The `Iterate` interface does not expose a way to have a cursor nor an iterator
+to the past-last element. An option would be to have `end()` return a predefined
+invalid value, to satisfy iterator comparison when `Next()` returns an empty
+`Optional`.
+
+Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
+need to copy and cache the last value, to allow for `*it` to be called. This
+puts a requirement on `ElementType` to be copyable for this interoperability to
+be usable.
+
+```cpp
+namespace Carbon {
+// A C++ input iterator for any type that implements the Carbon `Iterate` interface.
+template <typename Iterate>
+class IterateIterator {
+ public:
+  // Returns an "Invalid" iterator representing `end()`
+  static auto Invalid(const Iterate& container) -> IterateIterator<Iterate> {
+    return IterateIterator<Iterate>(container, std::nullopt);
+  }
+
+  IterateIterator(const Iterate& container,
+                  std::optional<typename Iterate::CursorType> cursor)
+      : container_(&container), cursor_(cursor) {
+    if (cursor_) {
+      next();
+    }
+  }
+  IterateIterator(const IterateIterator& other)
+      : container_(other.container_), cursor_(other.cursor_),
+      value_(other) {}
+
+  auto operator*() const -> const typename Iterate::ElementType& {
+    assert(cursor_);
+    return value_;
+  }
+  auto operator++() -> IterateIterator<Iterate>& {
+    next();
+    return *this;
+  }
+  auto operator==(const IterateIterator<Iterate>& rhs) -> bool {
+    return container_ == rhs.container_ && cursor_ == rhs.cursor_;
+  }
+  auto operator!=(const IterateIterator<Iterate>& rhs) -> bool {
+    return !(*this == rhs);
+  }
+
+ private:
+  void next() {
+    assert(cursor_);
+    if (const auto v = container_->Next(*cursor_); v) {
+      value_ = *v;
+    } else {
+      cursor_ = std::nullopt;
+    }
+  }
+
+  const Iterate* container_;
+  std::optional<typename Iterate::CursorType> cursor_;
+  typename Iterate::ElementType value_;
+};
+
+// A type implementing `Iterate` with CursorType `C`, and
+// ElementType `T` could have `begin()` and `end()` methods
+// defined as such:
+class IterateType ... {
+  auto begin() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
+  }
+  auto end() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>::Invalid(*this);
```

Comment:
What is `Iterable<...>` here? It looks like `IterateIterator` expects to be parameterized by a container type. Is `Iterable` some sort of adapter to define the `CursorType` and `ElementType` members? I don't see it defined anywhere else in this file.

Reply:
That's a typo, and should be  `Iterate<C, T>`, which is the C++ equivalent to Carbon's type. Fixed.